### PR TITLE
feature: reading namespaced ConfigMaps

### PR DESCRIPTION
**Pull request by aureamunoz**
Related to https://github.com/quarkusio/quarkus/pull/8376

### CODE REVIEW

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java [hunk 1]
```diff
@@ -28,4 +28,10 @@
     @ConfigItem
     public Optional<List<String>> configMaps;
 
+    /**
+     * Namespace to look for in
```
Comment (machi1990) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:32]:
```suggestion
     * Namespace to look for config maps
```

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java [hunk 2]
```diff
@@ -28,4 +28,10 @@
...
+     */
+    @ConfigItem
+    public String namespace;
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:35]: This needs to be `Optional`

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java [hunk 3]
```diff
@@ -49,12 +49,25 @@ public KubernetesConfigSourceProvider(KubernetesConfigSourceConfig config, Kuber
                 if (log.isDebugEnabled()) {
                     log.debug("Attempting to read ConfigMap " + configMapName);
                 }
-                ConfigMap configMap = client.configMaps().withName(configMapName).get();
+                ConfigMap configMap = client.configMaps().inNamespace(config.namespace).withName(configMapName).get();
                 if (configMap == null) {
-                    logMissingOrFail(configMapName, client.getNamespace(), "ConfigMap", config.failOnMissingConfig);
+                    log.debug("ConfigMap " + configMapName + " does not exist in configured namespace " + config.namespace
+                            + ". Trying to read it from the current workspace " + client.getNamespace());
+                    configMap = client.configMaps().withName(configMapName).get();
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java:56]: We probably don't want to do this, we would ideally just check if the Optional `namespace` property is set and act on way or the other

**Review by geoand — CHANGES_REQUESTED** [review]
Apart from the comments I added, I am not all that sure about this to be honest.

I can see the appeal, but on the other hand if we want more flexibility, we would allow users to individually configure the namespace of the configmap.

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java (hunk 2 above)
Comment (aureamunoz) [reply extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:35]: ok

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java (hunk 3 above)
Comment (aureamunoz) [reply extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java:56]: Sorry about the delay. So if `namespace` property is set we search for the configmap in that namespace and `logMissingOrFail` if it is not founded, right?

Comment (geoand) [reply extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java:56]: Yeah

**Comment by aureamunoz** [comment]
Could you please assign the issue to me, @geoand ?

**Comment by geoand** [comment]
Done

**Comment by aureamunoz** [comment]
Finally :sweat_smile: When you have some time @geoand I will be happy if you can review this. Thanks!

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java [hunk 4]
```diff
@@ -49,9 +49,17 @@ public KubernetesConfigSourceProvider(KubernetesConfigSourceConfig config, Kuber
                 if (log.isDebugEnabled()) {
                     log.debug("Attempting to read ConfigMap " + configMapName);
                 }
-                ConfigMap configMap = client.configMaps().withName(configMapName).get();
+                ConfigMap configMap = null;
+                String namespace = "";
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProvider.java:53]: Minor nitpick: I don't think you need to specify default values here

File: extensions/kubernetes-config/runtime/src/test/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                .addToData("some.key", "someValue").addToData("some.other", "someOtherValue").build();
+
+        KubernetesClient kubernetesClient = Mockito.mock(KubernetesClient.class);
+        stubNamespacedConfigMap(kubernetesClient, configMap, "cm1");
+
+        KubernetesConfigSourceProvider kcsp = new KubernetesConfigSourceProvider(config, kubernetesClient);
+        Iterable<ConfigSource> configSources = kcsp.getConfigSources(null);
+        assertThat(configSources).isNotEmpty();
+        ConfigSource next = configSources.iterator().next();
+        //        assertThat(next.getName()).isEqualTo("cm1");
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/test/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceProviderTest.java:54]: I assume you want to remove this?

File: integration-tests/kubernetes-client/src/main/resources/application.properties [hunk 6]
```diff
@@ -1,3 +1,4 @@
 quarkus.kubernetes-config.enabled=true
 quarkus.kubernetes-config.config-maps=cmap1,cmap2
 quarkus.kubernetes-config.secrets=s1
+quarkus.kubernetes-config.namespace=test
```
Comment (geoand) [inline integration-tests/kubernetes-client/src/main/resources/application.properties:4]: I am not sure I like the idea of having our main integration test use the less likely use case

File: integration-tests/kubernetes-client/src/test/java/io/quarkus/it/kubernetes/client/CustomKubernetesMockServerTestResource.java [hunk 7]
```diff
@@ -22,6 +26,24 @@ public void configureMockServer(KubernetesMockServer mockServer) {
...
+                        .addToData("dummy", "dummyFromDemo")
+                        .addToData("some.prop1", "val1FromDemo")
+                        .addToData("some.prop2", "val2FromDemo")
+                        .addToData("some.prop5", "val5FromDemo")
+                        .addToData("application.properties", "some.prop3=val3FromDemo")
+                        .addToData("application.yaml", "some:\n  prop4: val4FromDemo").build())
+                .once();
+    }
+
+    public static void assertProperty(String propertyName, String expectedValue) {
```
Comment (geoand) [inline integration-tests/kubernetes-client/src/test/java/io/quarkus/it/kubernetes/client/CustomKubernetesMockServerTestResource.java:41]: I don't think the assertions belong here

File: test-framework/kubernetes-client/src/main/java/io/quarkus/test/kubernetes/client/KubernetesMockServerTestResource.java [hunk 8]
```diff
@@ -23,6 +23,8 @@
 
         mockServer = new KubernetesMockServer(useHttps());
         mockServer.init();
+        System.out.println("[INFO] Kubernetes Mock server started; server status: " + mockServer.getHostName() + ":"
```
Comment (geoand) [inline test-framework/kubernetes-client/src/main/java/io/quarkus/test/kubernetes/client/KubernetesMockServerTestResource.java:26]: We probably don't want this

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java [hunk 9]
```diff
@@ -28,4 +28,10 @@
     @ConfigItem
     public Optional<List<String>> configMaps;
 
+    /**
+     * Namespace to look for config maps
+     */
+    @ConfigItem
+    public Optional<String> namespace;
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:35]: I am generally not super happy about having every configMap use the same namespace, but when I was writing the initial version of the extension I couldn't come up with an easy enough to use configuration that would cater both no namespace and different namespace configMaps, so I guess we can live with this.

File: integration-tests/kubernetes-client/src/test/java/io/quarkus/it/kubernetes/client/CustomKubernetesMockServerTestResource.java (hunk 7 above)
Comment (aureamunoz) [reply integration-tests/kubernetes-client/src/test/java/io/quarkus/it/kubernetes/client/CustomKubernetesMockServerTestResource.java:41]: yes, I moved to use it from other place and finally forgot undo the change.

File: test-framework/kubernetes-client/src/main/java/io/quarkus/test/kubernetes/client/KubernetesMockServerTestResource.java (hunk 8 above)
Comment (aureamunoz) [reply test-framework/kubernetes-client/src/main/java/io/quarkus/test/kubernetes/client/KubernetesMockServerTestResource.java:26]: true

**Comment by aureamunoz** [comment]
Thank you for the comments.

**Comment by geoand** [comment]
:welcome:

**Comment by geoand** [comment]
I'll review again later today

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java [hunk 10]
```diff
@@ -28,4 +28,10 @@
     @ConfigItem
     public Optional<List<String>> configMaps;
 
+    /**
+     * Namespace to look for config maps
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:32]: We should add more documentation on how this behaves

**Review by geoand — COMMENTED** [review]
I think that once we have the documentation issue I mentioned fixed, we can merge this

**Comment by aureamunoz** [comment]
I added some comments to the docs, let me know wdyt @geoand

File: extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java [hunk 11]
```diff
@@ -28,4 +28,11 @@
     @ConfigItem
     public Optional<List<String>> configMaps;
 
+    /**
+     * Namespace to look for config maps. Quarkus will return the configMaps in the namespace indicated by this property if they
+     * exists, build will fail otherwise.
```
Comment (geoand) [inline extensions/kubernetes-config/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesConfigSourceConfig.java:33]:
```suggestion
     * Namespace to look for config maps. If this is not specified, then the namespace configured in the kubectl config context is used. If the value is specified and the namespace doesn't exist, the application will fail to start.
```

File: docs/src/main/asciidoc/kubernetes-config.adoc [hunk 12]
```diff
@@ -40,8 +40,8 @@ The extension understands the following types of ConfigMaps as input sources:
...
 * ConfigMaps created from files named, `application.properties`, `application.yaml` or `application.yml` (see https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files[this] for an example on how to create one).
 
-To configure which ConfigMaps (from the namespace that the application runs in, or the explicitly configured namespace via `quarkus.kubernetes-client.namespace`), you can set the `quarkus.kubernetes-config.config-maps` property (in any of the usual Quarkus ways).
-Keep in mind however that you will also have to explicitly enable the retrieval of ConfigMaps by setting `quarkus.kubernetes-config.enabled=true` (the default is `false` in order to make it easy to test the application locally).
+To configure which ConfigMaps (from the namespace that the application runs in, or the explicitly configured namespace via `quarkus.kubernetes-client.namespace`), you can set the `quarkus.kubernetes-config.config-maps` property (in any of the usual Quarkus ways). If you want to access and read the configMaps from others namespaces you can set the
```
Comment (geoand) [inline docs/src/main/asciidoc/kubernetes-config.adoc:43]:
```suggestion
To configure which ConfigMaps (from the namespace that the application runs in, or the explicitly configured namespace via `quarkus.kubernetes-client.namespace`), you can set the `quarkus.kubernetes-config.config-maps` property (in any of the usual Quarkus ways). If access to ConfigMaps from a specific namespace you can set the
```

**Comment by aureamunoz** [comment]
Thank you @geoand

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by geoand** [comment]
The change might not pass formatting because I just added via Github suggestion

**Comment by aureamunoz** [comment]
Finally checks passed, I have just rebate onto master.

**Comment by aureamunoz** [comment]
I rebased onto master and Checks passed @geoand :-)

**Comment by geoand** [comment]
Thanks!

**Pull request merged**